### PR TITLE
Bt18 protag hybrids

### PR DESCRIPTION
**Pull request by Erdandrej**
Added all card effects of the Protagonist Hybrid Decks.

### CODE REVIEW

File: Assets/CardEffect/BT18/Black/AncientBeetlemon_BT18_072.cs [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+        {
+            List<ICardEffect> cardEffects = new List<ICardEffect>();
+
+            #region Alternate Digivolution
+
+            if (timing == EffectTiming.None)
+            {
+                bool PermanentCondition(Permanent targetPermanent)
+                {
+                    return targetPermanent.TopCard.IsLevel5 && targetPermanent.TopCard.ContainsTraits("Insectoid");
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/AncientBeetlemon_BT18_072.cs]: use targetPermanent.EqualsTrait instead

File: Assets/CardEffect/BT18/Black/Beetlemon_BT18_063.cs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            {
+                bool Condition()
+                {
+                    return card.Owner.HandCards.Contains(card);
+                }
+
+                bool PermanentCondition(Permanent targetPermanent)
+                {
+                    return targetPermanent.TopCard.EqualsCardName("J.P. Shibayama") ||
+                           targetPermanent.TopCard.EqualsCardName("J.P.Shibayama");
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/Beetlemon_BT18_063.cs]: When using EqualsCardName you do not need to include a check for no spaces. Method will do that for you

File: Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+                string EffectDescription()
+                {
+                    return
+                        "[Start of Your Main Phase] By trashing 1 card with the [Hybrid]/[Ten Warriors] trait in your hand, <Draw 1>.";
+                }
+
+                bool CanUseCondition(Hashtable hashtable)
+                {
+                    return CardEffectCommons.IsExistOnBattleArea(card) &&
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs:38]: use IsExistOnBattleAreaDigimon

File: Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                }
+
+                bool HasHybridInTrait(CardSource cardSource)
+                {
+                    return cardSource.HasHybridTenWarriorsTraits;
+                }
+
+                bool CanActivateCondition(Hashtable hashtable)
+                {
+                    return CardEffectCommons.IsExistOnBattleArea(card) &&
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs:49]: use IsExistOnBattleAreaDigimon

File: Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+                string EffectDescription()
+                {
+                    return
+                        "[All Turns] (Once Per Turn) When an attack target is changed, you may play 1 Tamer card with inherited effects from your hand without paying the cost.";
+                }
+
+                bool CanUseCondition(Hashtable hashtable)
+                {
+                    return CardEffectCommons.IsExistOnBattleArea(card) &&
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs:108]: IsExistOnBattleAreaDigimon

File: Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+                bool CanSelectCardCondition(CardSource cardSource)
+                {
+                    return cardSource.IsTamer && cardSource.HasInheritedEffect &&
+                           CardEffectCommons.CanPlayAsNewPermanent(cardSource: cardSource, payCost: false, cardEffect: activateClass);
+                }
+
+                bool CanActivateCondition(Hashtable hashtable)
+                {
+                    return CardEffectCommons.IsExistOnBattleArea(card) &&
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs:120]: IsExistOnBattleAreaDigimon

File: Assets/CardEffect/BT18/Black/KoKabuterimon_BT18_057.cs [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+                string EffectDescription()
+                {
+                    return
+                        "[Your Turn] (Once Per Turn) When this Digimon or any of your Tamers would digivolve into a multicolored Digimon that is black or yellow, reduce the digivolution cost by 1.";
+                }
+
+                bool CanUseCondition(Hashtable hashtable)
+                {
+                    return CardEffectCommons.IsExistOnBattleArea(card) &&
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/KoKabuterimon_BT18_057.cs]: IsExistOnBattleAreaDigimon

File: Assets/CardEffect/BT18/Black/KoKabuterimon_BT18_057.cs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+                bool DigivolveToCardCondition(CardSource cardSource)
+                {
+                    return cardSource.IsDigimon && cardSource.CardColors.Count > 1 &&
+                           (cardSource.CardColors.Contains(CardColor.Black) || cardSource.CardColors.Contains(CardColor.Yellow));
+                }
+
+                bool CanActivateCondition(Hashtable hashtable)
+                {
+                    return CardEffectCommons.IsExistOnBattleAreaDigimon(card);
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/KoKabuterimon_BT18_057.cs:50]: IsExistOnBattleAreaDigimon

File: Assets/CardEffect/BT18/Black/MetalKabuterimon_BT18_067.cs [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+            {
+                bool Condition()
+                {
+                    return card.Owner.HandCards.Contains(card);
+                }
+
+                bool PermanentCondition(Permanent targetPermanent)
+                {
+                    return targetPermanent.TopCard.EqualsCardName("J.P. Shibayama") ||
+                           targetPermanent.TopCard.EqualsCardName("J.P.Shibayama");
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/MetalKabuterimon_BT18_067.cs]: do not need to check for no spaces

File: Assets/CardEffect/BT18/Black/RhinoKabuterimon_BT18_070.cs [hunk 10]
```diff
@@ -0,0 +1,355 @@
...
+        {
+            List<ICardEffect> cardEffects = new List<ICardEffect>();
+
+            #region Alternate Digivolution
+
+            if (timing == EffectTiming.None)
+            {
+                bool PermanentCondition(Permanent targetPermanent)
+                {
+                    return targetPermanent.TopCard.IsLevel4 && targetPermanent.TopCard.ContainsTraits("Hybrid")
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/RhinoKabuterimon_BT18_070.cs]: use EqualsTrait

File: Assets/CardEffect/BT18/Black/RhinoKabuterimon_BT18_070.cs [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+            #region When Attacking - ESS
+
+            if (timing == EffectTiming.OnAllyAttack)
+            {
+                ActivateClass activateClass = new ActivateClass();
+                activateClass.SetUpICardEffect("DP -4000", CanUseCondition, card);
+                activateClass.SetUpActivateClass(CanActivateCondition, ActivateCoroutine, 1, false,
+                    EffectDescription());
+                activateClass.SetIsInheritedEffect(true);
+                cardEffects.Add(activateClass);
```
Comment (Michael8818) [inline Assets/CardEffect/BT18/Black/RhinoKabuterimon_BT18_070.cs]: needs SetHashstring

**Review by Michael8818 — CHANGES_REQUESTED** [review]

File: Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs (hunk 3 above)
Comment (Erdandrej) [reply Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs:38]: JP is a tamer. he'll never be in the battle area as a digimon

File: Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs (hunk 4 above)
Comment (Erdandrej) [reply Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs:49]: JP is a tamer. he'll never be in the battle area as a digimon

File: Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs (hunk 5 above)
Comment (Erdandrej) [reply Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs:108]: JP is a tamer. he'll never be in the battle area as a digimon

File: Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs (hunk 6 above)
Comment (Erdandrej) [reply Assets/CardEffect/BT18/Black/JPShibayama_BT18_091.cs:120]: JP is a tamer. he'll never be in the battle area as a digimon

**Review by Michael8818 — APPROVED** [review]

**Pull request merged**